### PR TITLE
Improved proofs

### PR DESCRIPTION
**Pull request by hjorthjort**
- Fix the styling of `loops-spec`, abstract it to any type.
- Split memory proof in two, one for concrete and one for abstract types
- Use the concrete memory proof to reintroudce some more things in the trusted computing base. More rules, but more obvious one's than the `#setRange(#getRange)` one.

### CODE REVIEW

File: data.md [hunk 1]
```diff
@@ -517,9 +517,9 @@ The function interprets the range of bytes as little-endian.
     syntax Int ::= #getRange (ByteMap, Int , Int) [function, functional]
  // --------------------------------------------------------------------
     rule #getRange(BM, START, WIDTH) => 0
-      requires notBool (WIDTH >Int 0)  [concrete]
+      requires notBool (WIDTH >Int 0)
```
Comment (ehildenb) [inline data.md]: these changes were needed to get it working with the Haskell backend (in an upcoming PR), so can't let this in.

Should be that if we go through a K submodule update, that the memory proofs work.

File: data.md [hunk 2]
```diff
@@ -517,9 +517,9 @@ The function interprets the range of bytes as little-endian.
...
     rule #getRange(BM, START, WIDTH) => #get(BM, START) +Int (#getRange(BM, START +Int 1, WIDTH -Int 1) *Int 256)
-      requires          WIDTH >Int 0   [concrete]
+      requires          WIDTH >Int 0
```
Comment (ehildenb) [inline data.md]: same here

File: data.md [hunk 3]
```diff
@@ -529,11 +529,11 @@ The function interprets the range of bytes as little-endian.
     syntax Int     ::= #get (ByteMap , Int     ) [function, smtlib(mapGet)]
     syntax ByteMap ::= #set (ByteMap , Int, Int) [function, smtlib(mapSet)]
  // -----------------------------------------------------------------------
-    rule #get( ByteMap <| M |>, KEY ) => {M [KEY]}:>Int requires         KEY in_keys(M) [concrete]
+    rule #get( ByteMap <| M |>, KEY ) => {M [KEY]}:>Int requires         KEY in_keys(M)
```
Comment (ehildenb) [inline data.md]: same here

File: data.md [hunk 4]
```diff
@@ -529,11 +529,11 @@ The function interprets the range of bytes as little-endian.
...
     rule #get( ByteMap <| M |>, KEY ) => 0              requires notBool KEY in_keys(M)
 
     rule #set( ByteMap <| M |>, KEY, VAL ) => ByteMap <| M [KEY <- undef] |> requires          VAL ==Int 0
-    rule #set( ByteMap <| M |>, KEY, VAL ) => ByteMap <| M [KEY <- VAL  ] |> requires notBool (VAL ==Int 0) [concrete]
+    rule #set( ByteMap <| M |>, KEY, VAL ) => ByteMap <| M [KEY <- VAL  ] |> requires notBool (VAL ==Int 0)
```
Comment (ehildenb) [inline data.md]: same here

File: kwasm-lemmas.md [hunk 5]
```diff
@@ -24,6 +24,17 @@ Basic arithmetic
       requires 0 <=Int X
        andBool X  <Int N
       [simplification]
+
+    rule (X *Int N +Int Y) modInt N => Y modInt N
+      [simplification]
+    rule (Y +Int X *Int N) modInt N => Y modInt N
+      [simplification]
```
Comment (ehildenb) [inline kwasm-lemmas.md]: can we have `[simplification]` on the same line as `rule` so they line up nicer?

File: tests/proofs/memory-abstract-type-lemmas.k [hunk 6]
```diff
@@ -0,0 +1,8 @@
+require "kwasm-lemmas.k"
+
+module MEMORY-ABSTRACT-TYPE-LEMMAS
+  imports KWASM-LEMMAS
+
+  rule #setRange(BM, EA, #getRange(BM, EA, WIDTH), WIDTH) => BM
+
+endmodule
```
Comment (ehildenb) [inline tests/proofs/memory-abstract-type-lemmas.k]: why is this moved to its own file?

File: tests/proofs/memory-concrete-type-spec.k [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+           <msize> SIZE </msize>
+           <mdata> BM   </mdata>
+           ...
+         </memInst>
+       requires #chop(<i32> ADDR) ==K <i32> EA
+        andBool EA +Int #numBytes(i64) <=Int SIZE *Int #pageSize()
+        andBool #isByteMap(BM)
+        andBool #inUnsignedRange(i32, ADDR)
+
+endmodule
```
Comment (ehildenb) [inline tests/proofs/memory-concrete-type-spec.k]: Isn't this just an instance of the abstract type proof? If so, are we having this one so it can run with the Haskell backend? In that case, let's try a submodule bump and see if it works with latest haskell backend. If not, let's notify the haskell backend team.

File: data.md (hunk 1 above)
Comment (hjorthjort) [reply data.md]: Good catch. I'll wait for the submodule update. But I think the concrete-type memory proofs won't go through, because in those `#getRange` and `#setRange` won't get expanded into `#get`.

File: tests/proofs/memory-abstract-type-lemmas.k (hunk 6 above)
Comment (hjorthjort) [reply tests/proofs/memory-abstract-type-lemmas.k]: I want to make sure the concrete-type memory proofs go through without this, because this makes them trivial. By having this applied only to the one proof the other rules in `kwasm-lemmas` get properly exercised.

File: tests/proofs/memory-concrete-type-spec.k (hunk 7 above)
Comment (hjorthjort) [reply tests/proofs/memory-concrete-type-spec.k]: I wanted the `kwasm-lemmas` to do more granular memory reasoning. With this setup, the above proof goes through without the `#setRange(#getRange)` lemma, just using modular arithmetic. This works becuse when the type is concrete, the `#getRange` and `#setRange` functions get expanded into a series of `#get` and `#set`, and the new lemmas can do reasoning over those.

Comment (hjorthjort) [reply tests/proofs/memory-concrete-type-spec.k]: I'll do the submodule bump and see if it works.

File: data.md (hunk 1 above)
Comment (hjorthjort) [reply data.md]: Is it the case that `[concrete]` forces a function to apply only when the entire term is concrete? Or can the annotation be made more granular? In this case, it would be nice if we force only the width to be concrete (given that it would still work with the Haskell backend).

File: tests/proofs/memory-abstract-type-lemmas.k (hunk 6 above)
Comment (hjorthjort) [reply tests/proofs/memory-abstract-type-lemmas.k]: Ah, you mean why not just a module in the same file? Good point, that's better.

File: Makefile [hunk 8]
```diff
@@ -145,11 +145,17 @@ TEST_CONCRETE_BACKEND       := llvm
 TEST_FLOAT_CONCRETE_BACKEND := java
 TEST_SYMBOLIC_BACKEND       := haskell
 
-tests/proofs/memory-spec.k.prove: TEST_SYMBOLIC_BACKEND=java
+tests/proofs/memory-concrete-type-spec.k.prove: TEST_SYMBOLIC_BACKEND=java
+tests/proofs/memory-concrete-type-spec.k.prove: PROVE_OPTIONS=--z3-tactic "(or-else (using-params smt :random-seed 3 :timeout 1000) (using-params smt :random-seed 2 :timeout 2000) (using-params smt :random-seed 0))" --z3-impl-timeout 6000
+tests/proofs/memory-concrete-type-spec.k.prove: KPROVE_MODULE=MEMORY-CONCRETE-TYPE-LEMMAS
+tests/proofs/memory-symbolic-type-spec.k.prove: TEST_SYMBOLIC_BACKEND=java
```
Comment (hjorthjort) [inline Makefile]: This needs to be removed before merging

File: tests/proofs/memory-concrete-type-spec.k (hunk 7 above)
Comment (hjorthjort) [reply tests/proofs/memory-concrete-type-spec.k]: So to exercise some more z3 level reasoning about memory I switched to `store16` and `load32_u` to make sure the `#setRange(#getRange)` lemma doesn't apply.

File: kwasm-lemmas.md [hunk 9]
```diff
@@ -70,21 +70,17 @@ We want to make the variant explicit, so we introduce the following helper, whic
...
+    rule #isByteMap(BMAP) impliesBool (0 <=Int #get(BMAP, IDX) andBool #get(BMAP, IDX) <Int 256) => true [smt-lemma]
 ```
 
 From the semantics, it should be clear that setting the index in a bytemap to the value already contained there will leave the map unchanged.
 Conversely, setting an index in a map to a value `VAL` and then retrieving the value at that index will yield `VAL`.
 
 ```k
-    rule #get(#set(BMAP, IDX, VAL), IDX)  => VAL
-    rule #set(BMAP, IDX, #get(BMAP, IDX)) => BMAP
+    rule #set(BMAP, IDX, #get(BMAP, IDX)) => BMAP [smt-lemma]
```
Comment (ehildenb) [inline kwasm-lemmas.md]: Let's simplify these ones in K directly.

File: data.md (hunk 1 above)
Comment (ehildenb) [reply data.md]: You can have more granular annotation with the `#isConcrete` meta-function, which you can plaace in the side condition and call on individual variables.
